File: data-fetcher/src/main/java/dataclasses/interfaces/Canteen.java
package dataclasses.interfaces;

import dataclasses.CanteenImplementation;
import dataclasses.Day;
import dataclasses.OpeningHoursImplementation;
import dataclasses.enums.Location;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public interface Canteen {
    String getName();
    Location getLocation();
    String getTitleInfo();
    List<OpeningHours> getOpeningHours();
    String getBodyInfo();
    String getLinkToFoodPlan();

    List<Day> getMenus();

    Optional<Day> getMenuOfDay(LocalDate date);

    static CanteenImplementation createCanteen(
            String name,
            Location location,
            String info,
            List<OpeningHours> openingHours,
            String additional,
            String linkToFoodPlan,
            List<Day> menusPerDay
    ) {
        return new CanteenImplementation(
                name,
                location,
                info,
                openingHours,
                additional,
                linkToFoodPlan,
                menusPerDay
        );
    }
}

File: src/main/kotlin/domain/model/Canteen.kt
package domain.model

import domain.model.enums.FoodProviderType
import domain.model.enums.Location
import domain.model.interfaces.FoodProvider
import java.time.LocalDate

data class Canteen(
    override val name: String,
    override val type: FoodProviderType,
    override val location: Location,
    override val titleInfo: String,
    override val bodyInfo: String,
    override val linkToFoodPlan: String,
    override val linkToMoreInformation: String,
    override val description: String,
    override val address: String,
    override val openingHours: List<OpeningInfo>,
    override val menus: List<Menu>
): FoodProvider {
    override fun getMenuOfDay(date: LocalDate): Result<Menu> {
        menus.find { menu -> menu.date == date }.also { foundMenu ->
            return if (foundMenu == null) {
                Result.failure(Exception("Menu not found!"))
            } else {
                Result.success(foundMenu)
            }
        }
    }
}
